agent/consul/servercert/manager.go

```diff
@@ -0,0 +1,261 @@
+// The server certificate is managed internally and used for peering control-plane traffic
+// to the TLS-enabled external gRPC port.
+type CertManager struct {
+	logger hclog.Logger
+
+	// config contains agent configuration necessary for the cert manager to operate.
+	config Config
+
+	// cache provides an API to issue internal RPC requests and receive notifications
+	// when there are changes.
+	cache Cache
+
+	// cacheUpdateCh receives notifications of cache update events for resources watched.
+	cacheUpdateCh chan cache.UpdateEvent
+
+	// store provides read-only access to the server's state store.
+	store Store
+
+	// tlsConfigurator receives the leaf cert and peering server name updates from the cert manager.
+	tlsConfigurator TLSConfigurator
+}
+
+func NewCertManager(deps Deps) *CertManager {
+	return &CertManager{
+		config:          deps.Config,
+		logger:          deps.Logger,
+		cache:           deps.Cache,
+		cacheUpdateCh:   make(chan cache.UpdateEvent, 1),
+		store:           deps.Store,
+		tlsConfigurator: deps.TLSConfigurator,
+	}
+}
+
+func (m *CertManager) Start(ctx context.Context) error {
+	if err := m.initializeWatches(ctx); err != nil {
+		return fmt.Errorf("failed to set up certificate watches: %w", err)
+	}
+	go m.handleUpdates(ctx)
+
+	m.logger.Info("initialized server certificate management")
+	return nil
+}
+
+func (m *CertManager) initializeWatches(ctx context.Context) error {
+	if m.config.ACLsEnabled {
+		// If ACLs are enabled we need to watch for server token updates and set/reset
+		// leaf cert updates as token updates arrive.
+		go m.watchServerToken(ctx)
+	} else {
+		// If ACLs are disabled we set up a single cache notification for leaf certs.
+		if err := m.watchLeafCert(ctx); err != nil {
+			return fmt.Errorf("failed to watch leaf: %w", err)
+		}
+	}
+	go m.watchCAConfig(ctx)
+
+	return nil
+}
+
+func (m *CertManager) watchServerToken(ctx context.Context) {
+	// We keep the last iteration's cancel function to reset watches.
+	var (
+		notifyCtx context.Context
+		cancel    context.CancelFunc = func() {}
+	)
+	retryLoopBackoff(ctx, func() error {
+		ws := memdb.NewWatchSet()
+		ws.Add(m.store.AbandonCh())
+		ws.Add(ctx.Done())
+
+		_, token, err := m.store.SystemMetadataGet(ws, structs.ServerManagementToken)
+		if err != nil {
+			return err
+		}
+		if token == nil {
+			m.logger.Debug("ACLs have not finished initializing")
+			return nil
+		}
+		if token.Value == "" {
+			// This should never happen. If the leader stored a token with this key it will not be empty.
+			return fmt.Errorf("empty token")
+		}
+		m.logger.Debug("server management token watch fired - resetting leaf cert watch")
+
+		// Cancel existing the leaf cert watch and spin up new one any time the server token changes.
+		// We need a valid token that the leader server recognizes, and when the
+		cancel()
+		notifyCtx, cancel = context.WithCancel(ctx)
```

Comment:
Do we want to be using `WithCancel` here? If parent `ctx` is cancelled the `cache.Notify` will be cancelled as well; IIUC we want `notifyCtx` to be completely independent of the parent `ctx` in this retry loop.